Different method of populating new tests

ewanharris opened the pull request:
This PRs moves away from the strange `.addon.test.js` syntax I added to allowing all files to be named `.test.js`. It does so by reading in all files in the Resources/ folder and filtering for `.test.js`. Then populating an ejs template based off this data. It also implements a blacklist to filter out any potential bad test files that could crash a certain platform.

I'm PRing to get some feedback on these points:
- Do people like this option?
- Should we have a main blacklist file in the scripts/top level folder, and then merge that with a blacklist file that gets copied from each repo to allow easier overriding i.e not having to copy the file from here.

sgtcoolguy commented on scripts/test.js:
```diff
@@ -411,7 +427,7 @@ function test(branch, platforms, target, deviceId, skipSdkInstall, cleanup, call
 	if (process.env.JENKINS || process.env.JENKINS_URL) {
 		tasks.push(function (next) {
-			cleanupModules(next);
+			cleanupModulesAndPlugins(next);
```
was the function also renamed? I only see the calls changed...

ewanharris replied:
Yeah function was renamed [here](https://github.com/appcelerator/titanium-mobile-mocha-suite/pull/33/files#diff-737e967a491eb091ac9d9d19c9078b0eR347)

sgtcoolguy commented on scripts/test.js:
```diff
@@ -384,6 +384,22 @@ function cleanupModules(next) {
+	const testFiles = walkSync(resourcesDir).filter(file => testFileRegex.test(file));
+
+	const appjs = ejs.render(fs.readFileSync(path.join(resourcesDir, 'app.js.ejs'), 'utf8'), { testFiles, blacklist: JSON.stringify(blacklist) });
+	fs.writeFileSync(path.join(PROJECT_DIR, 'Resources', 'app.js'), appjs);
```
why write this sync, if we are already setup to do async with a reference to the callback function?

sgtcoolguy replied:
path.join(resourcesDir, 'app.js')

ewanharris replied:
No reason other than preference. I'll change to non-sync method and pass the error into the callback.

sgtcoolguy commented on Resources/blacklist.js:
```diff
@@ -0,0 +1,18 @@
+
+	],
+	'windows': [
+
```
I assume that 'es6.rest.args.test.js' should be here? Or does it work now with latest JSC build for Windows?

ewanharris replied:
The updated JSC in 7 works fine with rest args

sgtcoolguy commented on Resources/app.js.ejs:
```diff
@@ -0,0 +1,134 @@
+}
+
+<% testFiles.forEach(function(testFile){ %>
+if (blacklist[platform].indexOf('<%- testFile %>') === -1) {
```
This feels a little odd to me, the if guard will be output into the app.js itself, and it seems strange to stick the full blacklist object in here too. I kind of assumed the filtering would occur in renderAppJS and would just output the matching test requires in here. I'd also just do the blacklist as JSON if it's just a simple mapping from platform to an array of strings, but maybe you could be more flexible and have a "blacklist.<platform>.js" file that would export a filtering function that could be combined with the `'.test.js'` filter done in `renderAppJS`.

I guess this is sort of dependent on whether we want the generated "app" to be treated as multi-platform or tailored to the platform we know we plan to test it with when we generate it...

Makes me think a couple more things:
- Should it then also avoid even copying the blacklisted files into the generated app?
- If we support dynamic picking up of tests, can we start partitioning os-specific tests into platform-specific folders in the suite too, to avoid even having say iOS suite test files in an android app? (I mean we already do guard with test filters, so maybe not a big deal but it'd be more idiomatic of Titanium apps).

ewanharris replied:
I agree, this isn't how I started off intending to implement it, I intended to keep the template as simple as possible. But the first time I ran it while developing I was reminded that the `-p\--platforms` option takes multiple os' so landed on this. Restricting to one platform would probably make the generated file nicer, and allow the two points you mentioned.

I went with using a `.js` file as I figured it'd be good to comment why it's blacklisted so people don't need to dive into a git history to figure out why.

ewanharris commented:
@sgtcoolguy Took another cycle on this and added the following

- Moved the app.js rendering into the same step as copying the mocha assets, we now only copy what's needed for a platform.
- Started splitting platform specific test in their own folders using the syntax `<platform_name>_tests`, this exposed appcelerator/titanium_mobile/pull/9735 which this pr is now dependent upon

ewanharris commented on Resources/blacklist.js:
```diff
@@ -0,0 +1,21 @@
+
+module.exports = {
+	'android': [
+		'ios_tests',
```
Could probably be smarter than hardcoding these

sgtcoolguy replied:
I think we should use the same default rules as the CLI build does which is to skip other platform named folders, to avoid a potential conflict with the CLI anyways (as in it's not in the blacklist but the CLI will strip the files anyways).

So in a hard-coded way, we should skip iphone/ios/windows folders on android; android/windows folders on iOS; etc - and the blacklist should ideally be some glob based syntax in addition, similar to entries in a .gitignore file's syntax.

I don't know if Windows still chokes on our ES6 test file or not. If not, then you might be able to punt on the blacklist glob stuff altogether for now until we actually need it again...

sgtcoolguy commented:
Silly question, but why do we need to use `android_tests`-style folders to achieve this? We already support using `Resources/<platform>` to "conditionally" include files in specific platform builds only. We could just put the ios/windows/android specific tests into folders like `Resources/android`, for example.

sgtcoolguy commented on Resources/app.js.ejs:
```diff
@@ -0,0 +1,131 @@
+	win,
+	$results = [],
+	failed = false,
+	should;;
```
remove extra semicolon, please

sgtcoolguy commented on scripts/test.js:
```diff
@@ -151,20 +151,53 @@ function addTiAppProperties(next) {
-	copyDir(path.join(SOURCE_DIR, 'Resources'), path.join(PROJECT_DIR, 'Resources'));
+function copyMochaAssetsAndRenderAppJs(platform, next) {
+
+	const blacklist = require(path.join(RESOURCES_DIR, 'blacklist.js'));
```
I assume the reason the blacklist lives under Resources is so that PRs to the SDK can modify the blacklist?

ewanharris replied:
Yeah that's the intention. Might it be sensible/easier to have a permanent one in the mocha suite (maybe under /scripts), and then copies in titanium_mobile/_windows that we merge into the primary one here?

sgtcoolguy commented on scripts/test.js:
```diff
@@ -151,20 +151,53 @@ function addTiAppProperties(next) {
+	fs.copySync(path.join(SOURCE_DIR, 'i18n'), path.join(PROJECT_DIR, 'i18n'));
+
+	const appjs = ejs.render(fs.readFileSync(path.join(RESOURCES_DIR, 'app.js.ejs'), 'utf8'), { testFiles });
+	fs.writeFileSync(path.join(PROJECT_DIR, 'Resources', 'app.js'), appjs);
```
use the async version and pass `next` as the callback?

sgtcoolguy commented on scripts/test.js:
```diff
@@ -344,14 +377,14 @@ function cleanNonGaSDKs(sdkPath, next) {
 			}
 			console.log('Removing ' + thisSDKPath);
-			wrench.rmdirRecursive(thisSDKPath, callback);
+			fs.removeSync(thisSDKPath, callback);
```
sync version doesn't take a callback argument. Use the `fs.remove` async version.

sgtcoolguy commented on scripts/test.js:
```diff
@@ -449,29 +482,30 @@ function test(branch, platforms, target, deviceId, skipSdkInstall, cleanup, call
+			if (err) {
+				return next(err);
+			}
+			let prefix;
```
```js
let prefix = platform;
if (target) {
  prefix += '.' + target;
}
```
